submission/distributor.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package submission
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/ctpolicy"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/loglist"
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+)
+
+const (
+	// RootsRefreshInterval is roots refresh interval.
+	RootsRefreshInterval = time.Hour * 24
+)
+
+// Distributor class operates policy-based submission across Logs.
+type Distributor struct {
+	ll *loglist.LogList
+
+	mu sync.RWMutex
+
+	// helper structs produced out of ll during init.
+	logClients map[string]client.AddLogClient
+	logRoots   loglist.LogRoots
+	rootPool   *ctfe.PEMCertPool
+
+	rootsRefreshTicker *time.Ticker
+
+	policy ctpolicy.CTPolicy
+}
+
+// Run starts regular roots updates.
+func (d *Distributor) run() {
+	d.rootsRefreshTicker = time.NewTicker(RootsRefreshInterval)
+	go func() {
+		for range d.rootsRefreshTicker.C {
+			d.refreshRoots(context.Background())
+		}
+	}()
+}
+
+// refreshRoots requests roots from Logs and updates local copy if necessary.
+func (d *Distributor) refreshRoots(ctx context.Context) {
+	// TODO(Mercurrent) add implementation.
+}
+
+// SubmitToLog is Submitter interface.
```

Comment:
What is the `Submitter` interface? I don't see it in this package.

Reply:
It's declared in races.go file same package. Updating description.

Reply:
Ah I see it now, didn't notice it half way through the file.

Reply:
It might be a good idea to move the `Submitter` interface to the top of races.go, now that it's being used in multiple files. Right now, it's buried after a bunch of unexported stuff.